File: backend/src/routes/experiments.ts
import type { FastifyInstance } from 'fastify';
import crypto from 'crypto';
import {
  saveExperiment,
  getExperimentsFiltered,
  searchExperiments,
  getExperimentStats,
  getExperiment,
} from '../db/experiments';
import { createRunHandle, emitEvent, spawnRun } from '../services/runner';
import type { ExperimentConfig, ExperimentRow } from '../types';
import { getApiSettings } from '../db/settings';

import type { RunHandle } from '../services/runner';

const runs = new Map<string, RunHandle>();

function rowToExperiment(row: ExperimentRow): Record<string, unknown> {
  return {
    id: row.id,
    slug: row.slug,
    n_agents: row.n_agents,
    n_ticks: row.n_ticks,
    persona_set: row.persona_set,
    status: row.status,
    started_at: row.started_at,
    finished_at: row.finished_at,
    elapsed_s: row.started_at
      ? Math.round(Date.now() / 1000 - new Date(row.started_at).getTime() / 1000)
      : 0,
    result_summary: row.result_summary ? JSON.parse(row.result_summary) : null,
  };
}

export default async function experimentsRoutes(app: FastifyInstance) {
  app.get('', async (req) => {
    const { status, slug, limit = '20', offset = '0' } = req.query as Record<string, string>;
    const { rows, total } = getExperimentsFiltered(
      status || undefined,
      slug || undefined,
      parseInt(limit, 10) || 20,
      parseInt(offset, 10) || 0,
    );
    return {
      experiments: rows.map(rowToExperiment),
      total,
      limit: parseInt(limit, 10) || 20,
      offset: parseInt(offset, 10) || 0,
    };
  });

  app.get('/search', async (req) => {
    const { q = '', limit = '20' } = req.query as Record<string, string>;
    if (!q) return { experiments: [] };
    const rows = searchExperiments(q, parseInt(limit, 10) || 20);
    return { experiments: rows.map(rowToExperiment) };
  });

  app.get('/stats', async () => {
    return getExperimentStats();
  });

  app.get('/:expId', async (req, reply) => {
    const { expId } = req.params as { expId: string };
    const row = getExperiment(expId);
    if (row) {
      return { experiment: rowToExperiment(row) };
    }
    const handle = runs.get(expId);
    if (handle) {
      return {
        experiment: {
          id: handle.runId,
          slug: handle.slug,
          n_agents: handle.nAgents,
          n_ticks: handle.nTicks,
          persona_set: handle.personaSet,
          status: handle.finished ? 'completed' : 'running',
          started_at: new Date(handle.startedAt * 1000).toISOString(),
          finished_at: handle.finished ? new Date().toISOString() : null,
          elapsed_s: Math.round(Date.now() / 1000 - handle.startedAt),
          result_summary: null,
        },
      };
    }
    reply.status(404);
    return { message: 'Experiment not found' };
  });

  app.post('', async (req) => {
    const body = req.body as ExperimentConfig;
    const runId = crypto.randomBytes(12).toString('hex').slice(0, 12);
    const handle = createRunHandle(
      runId,
      body.slug,
      body.n_agents,
      body.n_ticks,
      body.persona_set,
    );
    runs.set(runId, handle);

    saveExperiment({
      id: runId,
      slug: body.slug,
      n_agents: body.n_agents,
      n_ticks: body.n_ticks,
      persona_set: body.persona_set,
      status: 'running',
      started_at: new Date().toISOString(),
      finished_at: null,
      result_summary: null,
    });

    const settings = getApiSettings();
    const apiSettings = settings
      ? { api_key: settings.api_key, base_url: settings.base_url, model: settings.model }
      : undefined;

    spawnRun(
      handle,
      (kind: string, data: Record<string, unknown>) => {
        emitEvent(handle, kind, data);
        if (kind === '__end__') {
          const metrics = handle.finalMetrics;
          const payload: Partial<ExperimentRow> = {
            id: runId,
            finished_at: new Date().toISOString(),
            result_summary: metrics ? JSON.stringify(metrics) : null,
            final_yes_mid: metrics.yes_mid_final as number | undefined,
            total_fills: metrics.n_fills as number | undefined,
            total_actions: metrics.n_actions as number | undefined,
            avg_tick_time_ms: handle.tickCount
              ? parseFloat(((handle.tickElapsedTotal / handle.tickCount) * 1000).toFixed(2))
              : undefined,
          };
          if (!handle.cancel) {
            payload.status = 'completed';
          }
          saveExperiment(payload);
        }
      },
      apiSettings,
    );

    return { run_id: runId };
  });

  app.post('/:expId/cancel', async (req, reply) => {
    const { expId } = req.params as { expId: string };
    const handle = runs.get(expId);
    if (!handle) {
      reply.status(404);
      return { message: 'Experiment not found' };
    }
    handle.cancel = true;
    saveExperiment({
      id: expId,
      status: 'cancelled',
      finished_at: new Date().toISOString(),
    });
    return { cancelled: true };
  });

  app.get('/:expId/events', async (req, reply) => {
    const { expId } = req.params as { expId: string };
    const { replay = '1' } = req.query as Record<string, string>;
    const wantReplay = replay !== '0' && replay !== 'false';

    const handle = runs.get(expId);
    if (handle) {
      // Live or recently finished run — stream from in-memory queue
      reply.raw.writeHead(200, {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive',
      });

      if (wantReplay) {
        for (const ev of handle.history) {
          if (ev.kind === '__end__') continue;
          reply.raw.write(`event: ${ev.kind}\ndata: ${JSON.stringify(ev.data)}\n\n`);
        }
      }

      let idx = handle.history.length;
      const timer = setInterval(() => {
        while (idx < handle.queue.length) {
          const ev = handle.queue[idx++];
          if (ev.kind === '__end__') {
            reply.raw.write(`event: end\ndata: {}\n\n`);
            clearInterval(timer);
            reply.raw.end();
            return;
          }
          reply.raw.write(`event: ${ev.kind}\ndata: ${JSON.stringify(ev.data)}\n\n`);
        }
        if (handle.finished && idx >= handle.queue.length) {
          reply.raw.write(`event: end\ndata: {}\n\n`);
          clearInterval(timer);
          reply.raw.end();
          return;
        }
        reply.raw.write(`event: ping\ndata: {}\n\n`);
      }, 250);

      req.raw.on('close', () => {
        clearInterval(timer);
      });

      return reply;
    }

    // Fallback: check SQLite for completed experiments
    const row = getExperiment(expId);
    if (row && row.result_summary) {
      reply.raw.writeHead(200, {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive',
      });
      try {
        const summary = JSON.parse(row.result_summary);
        reply.raw.write(`event: settled\ndata: ${JSON.stringify(summary)}\n\n`);
      } catch {
        reply.raw.write(`event: settled\ndata: {}\n\n`);
      }
      reply.raw.write(`event: end\ndata: {}\n\n`);
      reply.raw.end();
      return reply;
    }

    reply.status(404);
    return { message: 'Experiment not found' };
  });
}

File: legacy/webapp/backend/routers/experiments.py
"""Experiments API router."""
from __future__ import annotations

import asyncio
import json
import queue
import threading
import time
import uuid
from dataclasses import dataclass, field

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import JSONResponse, StreamingResponse
from sse_starlette.sse import EventSourceResponse

from database import (
    get_experiments, get_experiment, save_experiment,
    get_experiments_filtered, search_experiments as search_experiments_db,
    get_experiment_stats,
)
from models.experiment import (
    ExperimentConfig, CreateExperimentResponse,
    CancelExperimentResponse,
)

router = APIRouter(prefix="/experiments", tags=["experiments"])

# -------------------------------------------------------------------
# Run registry — keeps a per-run event queue + cancellation flag
# -------------------------------------------------------------------

@dataclass
class RunHandle:
    run_id: str
    slug: str
    n_agents: int
    n_ticks: int
    persona_set: str
    queue: "queue.Queue[tuple[str, dict]]" = field(default_factory=queue.Queue)
    cancel: threading.Event = field(default_factory=threading.Event)
    started_at: float = field(default_factory=time.time)
    finished: bool = False
    history: list[tuple[str, dict]] = field(default_factory=list)
    final_metrics: dict = field(default_factory=dict)
    tick_elapsed_s_total: float = 0.0
    tick_count: int = 0

_RUNS: dict[str, RunHandle] = {}
_RUNS_LOCK = threading.Lock()
_HISTORY_CAP = 2000


def _make_emitter(handle: RunHandle):
    def emit(kind: str, data: dict) -> None:
        handle.queue.put((kind, data))
        if len(handle.history) < _HISTORY_CAP:
            handle.history.append((kind, data))
        if kind == "settled":
            handle.final_metrics = data
        elif kind == "tick_finished":
            handle.tick_elapsed_s_total += data.get("elapsed_s", 0)
            handle.tick_count += 1
    return emit


def _spawn_run(handle: RunHandle) -> None:
    emit = _make_emitter(handle)

    def _worker():
        crashed = False
        try:
            # Import runner here to avoid startup cost
            from webapp.runner_stream import run_stream
            run_stream(
                slug=handle.slug,
                n_agents=handle.n_agents,
                n_ticks_override=handle.n_ticks,
                persona_set=handle.persona_set,
                on_event=emit,
                cancel=handle.cancel,
            )
        except Exception as exc:
            crashed = True
            logging = __import__("logging")
            logging.getLogger(__name__).exception("run %s crashed", handle.run_id)
            emit("error", {"where": "worker", "message": str(exc)})
        finally:
            handle.finished = True
            emit("__end__", {})
            metrics = handle.final_metrics
            payload: dict = {
                "id": handle.run_id,
                "finished_at": time.time(),
                "result_summary": json.dumps(metrics) if metrics else None,
                "final_yes_mid": metrics.get("yes_mid_final"),
                "total_fills": metrics.get("n_fills"),
                "total_actions": metrics.get("n_actions"),
                "avg_tick_time_ms": round(
                    handle.tick_elapsed_s_total / handle.tick_count * 1000, 2
                ) if handle.tick_count else None,
            }
            if crashed:
                payload["status"] = "error"
            elif not handle.cancel.is_set():
                payload["status"] = "completed"
            save_experiment(payload)

    threading.Thread(target=_worker, name=f"run-{handle.run_id[:8]}",
                     daemon=True).start()


# -------------------------------------------------------------------
# Endpoints
# -------------------------------------------------------------------

@router.get("")
def list_experiments(
    status: str | None = None,
    slug: str | None = None,
    limit: int = 20,
    offset: int = 0,
):
    exps, total = get_experiments_filtered(status=status, slug=slug, limit=limit, offset=offset)
    return {"experiments": exps, "total": total, "limit": limit, "offset": offset}


@router.get("/search")
def search_experiments(q: str = "", limit: int = 20):
    if not q:
        return {"experiments": []}
    exps = search_experiments_db(q=q, limit=limit)
    return {"experiments": exps}


@router.get("/stats")
def experiments_stats():
    return get_experiment_stats()


@router.get("/{exp_id}")
def read_experiment(exp_id: str):
    exp = get_experiment(exp_id)
    if exp is None:
        # Fallback to in-memory
        handle = _RUNS.get(exp_id)
        if handle:
            exp = {
                "id": handle.run_id,
                "slug": handle.slug,
                "n_agents": handle.n_agents,
                "n_ticks": handle.n_ticks,
                "persona_set": handle.persona_set,
                "status": "running" if not handle.finished else "completed",
                "started_at": handle.started_at,
                "finished_at": time.time() if handle.finished else None,
                "elapsed_s": round(time.time() - handle.started_at, 1),
                "result_summary": None,
            }
        else:
            raise HTTPException(404, "Experiment not found")
    return {"experiment": exp}


@router.post("", response_model=CreateExperimentResponse)
def create_experiment(config: ExperimentConfig):
    run_id = uuid.uuid4().hex[:12]
    handle = RunHandle(
        run_id=run_id, slug=config.slug,
        n_agents=config.n_agents, n_ticks=config.n_ticks,
        persona_set=config.persona_set,
    )
    with _RUNS_LOCK:
        _RUNS[run_id] = handle

    # Persist to SQLite
    save_experiment({
        "id": run_id,
        "slug": config.slug,
        "n_agents": config.n_agents,
        "n_ticks": config.n_ticks,
        "persona_set": config.persona_set,
        "status": "running",
        "started_at": time.time(),
        "finished_at": None,
        "elapsed_s": 0,
        "result_summary": None,
    })

    _spawn_run(handle)
    return {"run_id": run_id}


@router.post("/{exp_id}/cancel", response_model=CancelExperimentResponse)
def cancel_experiment(exp_id: str):
    handle = _RUNS.get(exp_id)
    if handle is None:
        raise HTTPException(404, "Experiment not found")
    handle.cancel.set()
    save_experiment({
        "id": exp_id,
        "status": "cancelled",
        "finished_at": time.time(),
    })
    return {"cancelled": True}


@router.get("/{exp_id}/events")
async def stream_events(exp_id: str, request: Request, replay: int = 1):
    handle = _RUNS.get(exp_id)
    if handle is None:
        raise HTTPException(404, "Experiment not found")

    async def generator():
        # Replay buffered history first
        if replay:
            for kind, data in list(handle.history):
                if kind == "__end__":
                    continue
                yield {"event": kind, "data": json.dumps(data, default=str)}

        loop = asyncio.get_event_loop()
        while True:
            if await request.is_disconnected():
                break
            try:
                kind, data = await loop.run_in_executor(
                    None, lambda: handle.queue.get(timeout=15.0),
                )
            except queue.Empty:
                yield {"event": "ping", "data": "{}"}
                if handle.finished:
                    break
                continue
            if kind == "__end__":
                yield {"event": "end", "data": "{}"}
                break
            yield {"event": kind, "data": json.dumps(data, default=str)}

    return EventSourceResponse(generator())
